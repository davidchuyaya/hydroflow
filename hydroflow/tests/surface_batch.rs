use hydroflow::util::collect_ready;
use hydroflow::{assert_graphvis_snapshots, hydroflow_syntax};
use multiplatform_test::multiplatform_test;

#[multiplatform_test]
pub fn test_basic_2() {
    let (signal_tx, signal_rx) = hydroflow::util::unbounded_channel::<()>();
    let (egress_tx, mut egress_rx) = hydroflow::util::unbounded_channel();

    let mut df = hydroflow_syntax! {
        gate = defer_signal::<'static>();
        source_iter([1, 2, 3]) -> [input]gate;
        source_stream(signal_rx) -> [signal]gate;

        gate -> for_each(|x| egress_tx.send(x).unwrap());
    };
    assert_graphvis_snapshots!(df);

    df.run_available();
    let out: Vec<_> = collect_ready(&mut egress_rx);
    assert_eq!(out, [0; 0]);

    signal_tx.send(()).unwrap();
    df.run_available();

    let out: Vec<_> = collect_ready(&mut egress_rx);
    assert_eq!(out, vec![1, 2, 3]);
}

#[multiplatform_test(test, env_tracing)]
pub fn test_basic_2_tick() {
    let (signal_tx, signal_rx) = hydroflow::util::unbounded_channel::<()>();
    let (egress_tx, mut egress_rx) = hydroflow::util::unbounded_channel();

    let mut df = hydroflow_syntax! {
        gate = defer_signal::<'tick>();
        source_iter([1, 2, 3]) -> fold::<'tick>(|| 0, |acc, _v| *acc += 1) -> [input]gate;
        source_stream(signal_rx) -> [signal]gate;

        gate -> for_each(|x| egress_tx.send(x).unwrap());
    };
    assert_graphvis_snapshots!(df);

    df.run_available();
    let out: Vec<_> = collect_ready(&mut egress_rx);
    assert_eq!(out, [0; 0]);

    signal_tx.send(()).unwrap();
    println!("2nd run available");
    df.run_available();

    let out: Vec<_> = collect_ready(&mut egress_rx);
    assert_eq!(out, vec![0]);

    df.run_available();

    let out: Vec<_> = collect_ready(&mut egress_rx);
    assert_eq!(out, [0; 0]);
}
