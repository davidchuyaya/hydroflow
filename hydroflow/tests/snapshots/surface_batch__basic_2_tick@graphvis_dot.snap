---
source: hydroflow/tests/surface_batch.rs
expression: "df.meta_graph().unwrap().to_dot(& Default :: default())"
---
digraph {
    node [fontname="Monaco,Menlo,Consolas,&quot;Droid Sans Mono&quot;,Inconsolata,&quot;Courier New&quot;,monospace", style=filled];
    edge [fontname="Monaco,Menlo,Consolas,&quot;Droid Sans Mono&quot;,Inconsolata,&quot;Courier New&quot;,monospace"];
    n1v1 [label="(n1v1) defer_signal::<'tick>()", shape=invhouse, fillcolor="#88aaff"]
    n2v1 [label="(n2v1) source_iter([1, 2, 3])", shape=invhouse, fillcolor="#88aaff"]
    n3v1 [label="(n3v1) fold::<'tick>(|| 0, |acc, _v| *acc += 1)", shape=invhouse, fillcolor="#88aaff"]
    n4v1 [label="(n4v1) source_stream(signal_rx)", shape=invhouse, fillcolor="#88aaff"]
    n5v1 [label="(n5v1) for_each(|x| egress_tx.send(x).unwrap())", shape=house, fillcolor="#ffff88"]
    n6v1 [label="(n6v1) handoff", shape=parallelogram, fillcolor="#ddddff"]
    n7v1 [label="(n7v1) handoff", shape=parallelogram, fillcolor="#ddddff"]
    n8v1 [label="(n8v1) handoff", shape=parallelogram, fillcolor="#ddddff"]
    n3v1 -> n6v1
    n2v1 -> n7v1
    n4v1 -> n8v1
    n1v1 -> n5v1
    n6v1 -> n1v1 [label="input", color=red]
    n7v1 -> n3v1 [color=red]
    n8v1 -> n1v1 [label="signal", color=red]
    subgraph "cluster n1v1" {
        fillcolor="#dddddd"
        style=filled
        label = "sg_1v1\nstratum 0"
        n2v1
    }
    subgraph "cluster n2v1" {
        fillcolor="#dddddd"
        style=filled
        label = "sg_2v1\nstratum 1"
        n3v1
    }
    subgraph "cluster n3v1" {
        fillcolor="#dddddd"
        style=filled
        label = "sg_3v1\nstratum 0"
        n4v1
    }
    subgraph "cluster n4v1" {
        fillcolor="#dddddd"
        style=filled
        label = "sg_4v1\nstratum 2"
        n1v1
        n5v1
        subgraph "cluster_sg_4v1_var_gate" {
            label="var gate"
            n1v1
        }
    }
}

